### PR TITLE
Docker 20.10.0 Major Update

### PR DESCRIPTION
**Pull request by G-M0N3Y-2503**
Maintainer: @G-M0N3Y-2503
Compile tested: x86_x64, Hyper-V, OpenWrt Master
Run tested: x86_x64, Hyper-V, OpenWrt Master

### Description
* docker/dockerd 20.10.0 Update
* SELinux and Seccomp are now enabled via the kernel options themselves instead of a package configuration option
* Config.in now uses CGroupsV2 by default but optionally allows CGroupsV1 to be enabled This will allow a "useable" package without configuring the kernel differently.

### CODE REVIEW

**Comment by jmarcet** [comment]
I can confirm docker-compose works all right with this update, only need to replace the dep in its Makefile from __docker-ce__ to __docker__.

I also confirm everything seems to be working with only CGroupsV2 enabled; without the cgroupfs-mount package and **cgroup_no_v1=all** added to /proc/cmdline

**Comment by dangowrt** [comment]
> I also confirm everything seems to be working with only CGroupsV2 enabled; without the cgroupfs-mount package and **cgroup_no_v1=all** added to /proc/cmdline

Beautiful! I was thinking of making `cgroup_no_v1=all` the default, ie. patching kernel to change default rather than having to append it to cmdline but still allowing people to divert from that by actually setting `cgroup_no_v1=none` or stuff like that:
```diff
--- a/kernel/cgroup/cgroup-v1.c
+++ b/kernel/cgroup/cgroup-v1.c
@@ -29,7 +29,7 @@
 #define CGROUP_PIDLIST_DESTROY_DELAY	HZ
 
 /* Controllers blocked by the commandline in v1 */
-static u16 cgroup_no_v1_mask;
+static u16 cgroup_no_v1_mask = U16_MAX;
 
 /* disable named v1 mounts */
 static bool cgroup_no_v1_named;
@@ -1252,6 +1252,8 @@ static int __init cgroup_no_v1(char *str
 	char *token;
 	int i;
 
+	cgroup_no_v1_mask = 0;
+
 	while ((token = strsep(&str, ",")) != NULL) {
 		if (!*token)
 			continue;
```

It'd also be great if more people would try using `uxc` instead of `runc` and report any missing features or bugs they encounter. I'm aware that console/terminal support is not implemented in a way that can work with docker/podman/... I might do that at some point during this week. But apart from that things should work with 98% features implemented at 2% of space and memory resource usage compared to `runc` (and still a small fraction of `crun`)

**Comment by G-M0N3Y-2503** [comment]
> It'd also be great if more people would try using `uxc` instead of `runc` and report any missing features or bugs they encounter.

In terms of docker, do you mean something like patching [containerd](https://github.com/containerd/containerd) to use uxc instead of runc, or uxc by itself?

**Comment by jmarcet** [comment]
I forgot to mention luci-app-dockerman works all right too with the same minimal change to its Makefile.

**Comment by feckert** [comment]
@jmarcet Could you please open also a P/R in the luci feed as a draft for luci-app-dockerman?
So all changes are staged which is related to this change.

**Comment by jmarcet** [comment]
> @jmarcet Could you please open also a P/R in the luci feed as a draft for luci-app-dockerman?
> So all changes are staged which is related to this change.

[#4590](https://github.com/openwrt/luci/pull/4590)

**Comment by dangowrt** [comment]
I've cherry-picked https://github.com/openwrt/packages/pull/13874/commits/4e603d21abd80de03ded8ae8568794a6f68f5201 onto https://github.com/openwrt/packages/commit/3a06ce55959016653767a1d2c276594ad1eeb743 (with some minor corrections).

File: utils/containerd/Makefile [hunk 1]
```diff
@@ -53,8 +42,8 @@ MAKE_FLAGS += \
 	VERSION=$(PKG_VERSION) \
 	REVISION=$(PKG_SOURCE_VERSION)
 
-ifeq ($(CONFIG_CONTAINERD_SECCOMP),y)
-MAKE_FLAGS += BUILDTAGS='seccomp'
```
Comment (dangowrt) [inline utils/containerd/Makefile:57]: Here you remove the check for the no longer existing `CONFIG_CONTAINERD_SECCOMP`. But where is the new check, eg. for `CONFIG_KERNEL_SECCOMP_FILTER`?

**Review by dangowrt — CHANGES_REQUESTED** [review]
Typo in the commit message `docker-ce: docker-ce -> dockerd`:
```
The source is being depricated and ...
                    ^^^^^^^^^^
should be "deprecated"
```
Also, see comment about `seccomp` flag not being added.

File: utils/containerd/Makefile (hunk 1 above)
Comment (G-M0N3Y-2503) [reply utils/containerd/Makefile:57]: [It seems that tag isn't useful anymore](https://github.com/containerd/containerd/blob/v1.4.1/releases/v1.4.0.toml#L62).
however, I'll remove `libseccomp` as I seemed to have missed it.

**Review by dangowrt — APPROVED** [review]
Looks ripe for merge to me

File: utils/dockerd/Config.in [hunk 2]
```diff
@@ -1,95 +1,83 @@
...
-config DOCKER_SECCOMP
-	bool "Enable support for seccomp in Docker"
-	depends on PACKAGE_docker-ce
+config DOCKER_GENERALLY_NECESSARY
+	bool "Enable generally necessary kernel support for Docker"
 	default n
-	select KERNEL_SECCOMP
-	select PACKAGE_libseccomp
+	depends on PACKAGE_dockerd
+	select KERNEL_KEYS
```
Comment (dangowrt) [inline utils/dockerd/Config.in:19]: Selecting `CONFIG_KEYS` by default for target kernels with large enough flash space might generally be a good idea.
I've been preparing a cleanup of the kernel keyring options, please see my staging tree:
https://git.openwrt.org/?p=openwrt/staging/dangole.git;a=commit;h=308aee3f80cbb91b2ddae67a208062f36b42c16a

Comment (G-M0N3Y-2503) [reply utils/dockerd/Config.in:19]: Yeah, that sounds pretty good.

Comment (dangowrt) [reply utils/dockerd/Config.in:19]: I've posted those changes to the mailing list, if there are no objections, I'll merge them this week:
https://patchwork.ozlabs.org/project/openwrt/patch/X9U+Vn+oBpAx3H+y@makrotopia.org/

Comment (dangowrt) [reply utils/dockerd/Config.in:19]: Pushed: https://github.com/openwrt/openwrt/commit/23049f9c31f030829980b7e32e424435572ae9ac
Please adapt the PR accordingly, other than that it's ready to get merged.

Comment (G-M0N3Y-2503) [reply utils/dockerd/Config.in]: Removed config option from dockerd

**Pull request merged**